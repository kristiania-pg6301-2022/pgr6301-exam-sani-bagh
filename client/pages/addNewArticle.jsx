import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import { FormInput } from "../lib/formInput";
import { NewsApiContext } from "../newsApiContext";

export function AddNewArticle() {
  const { createArticle } = useContext(NewsApiContext);
  const [title, setTitle] = useState("");
  const [author, setAuthor] = useState("");
  const [topic, setTopic] = useState("");
  const [text, setText] = useState("");

  const navigate = useNavigate();

  function handleSubmit(event) {
    event.preventDefault();
    createArticle({ title, author, topic, text });
    navigate("/");
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>Add new article</h1>

      <FormInput label={"Title:"} value={title} onChangeValue={setTitle} />
      <FormInput label={"Author:"} value={author} onChangeValue={setAuthor} />
      <FormInput label={"Topic:"} value={topic} onChangeValue={setTopic} />
      <FormInput label={"Text:"} value={text} onChangeValue={setText} />
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
