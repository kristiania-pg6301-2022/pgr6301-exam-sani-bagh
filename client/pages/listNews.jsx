import React, { useContext, useState } from "react";
import { useLoading } from "../useLoading";
import { NewsApiContext } from "../newsApiContext";
import { fetchJSON } from "../lib/fetchJSON";

function ArticleCard({ article: { title, author, category, text } }) {
  return (
    <>
      <h3>{title}</h3>
      <h3>Category: {category}</h3>
      <div>{author}</div>
      <p>{text}</p>
    </>
  );
}

export function ListNews() {
  const { listNews } = useContext(NewsApiContext);
  const [category, setCategory] = useState("");
  const [categoryQuery, setCategoryQuery] = useState("");
  const { loading, error, data } = useLoading(
    async () => await listNews({ category }),
    [category]
  );

  function handleSubmitQuery(e) {
    e.preventDefault();
    setCategory(categoryQuery);
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>News in the database</h1>

      <div>
        <form onSubmit={handleSubmitQuery}>
          <label>
            Topic:
            <input
              id="category-query"
              value={categoryQuery}
              onChange={(e) => setCategoryQuery(e.target.value)}
            />
            <button>Filter</button>
          </label>
        </form>
      </div>

      {data.map((article, index) => (
        <ArticleCard key={index} article={article} />
      ))}
    </div>
  );
}

export function ListAllNews() {
  const { listAllNews } = useContext(NewsApiContext);
  const [category, setCategory] = useState("");
  const { loading, error, data } = useLoading(
    async () => await listAllNews({ category }),
    [category]
  );

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <div>
      <h1>Daily posted Articles</h1>
      <ul>
        {data.map((article, index) => (
          <ArticleCard key={index} article={article} />
        ))}
      </ul>
    </div>
  );
}
