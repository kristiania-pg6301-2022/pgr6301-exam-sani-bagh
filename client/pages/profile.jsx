import { fetchJSON } from "../lib/fetchJSON";
import { useLoading } from "../useLoading";

export function Profile() {
  const { loading, data, error } = useLoading(async () => {
    return await fetchJSON("/api/login");
  });

  if (loading) {
    return <div>Please wait...</div>;
  }
  if (error) {
    return <div>Error! {error.toString()}</div>;
  }

  return (
    <div>
      <h1>
        Profile for {data.name || data.pid}
        {data.email && <span>({data.email})</span>}
      </h1>
      {data.picture && (
        <div>
          <img src={data.picture} alt={"Profile picture"} />
        </div>
      )}
    </div>
  );
}
