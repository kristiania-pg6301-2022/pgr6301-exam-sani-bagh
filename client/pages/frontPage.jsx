import { Link } from "react-router-dom";
import React from "react";
import { ListAllNews } from "./listNews";

export function FrontPage() {
  return (
    <div>
      <h1>Daily News</h1>
      <ListAllNews />
    </div>
  );
}
