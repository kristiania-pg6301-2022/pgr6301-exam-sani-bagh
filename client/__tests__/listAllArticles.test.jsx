import React from "react";
import ReactDOM from "react-dom";
import { ListAllNews, ListNews } from "../pages/listNews";
import { act } from "react-dom/test-utils";
import { NewsApiContext } from "../newsApiContext";

describe("ListAllNews component", () => {
  it("should show loading screen", function () {
    const domElement = document.createElement("div");
    ReactDOM.render(<ListAllNews />, domElement);
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("should show error message", async function () {
    const domElement = document.createElement("div");
    await act(async () => {
      const listAllNews = () => {
        throw new Error("Failed to load");
      };
      ReactDOM.render(
        <NewsApiContext.Provider value={{ listAllNews }}>
          <ListAllNews />
        </NewsApiContext.Provider>,
        domElement
      );
    });

    expect(domElement.querySelector("#error-text").innerHTML).toEqual(
      "Error: Failed to load"
    );
    expect(domElement.innerHTML).toMatchSnapshot();
  });

  it("should list all news", async function () {
    const articles = [
      { title: "Article 1", category: "Topic 1" },
      { title: "Article 2", category: "Topic 2" },
    ];
    const domElement = document.createElement("div");
    await act(async () => {
      ReactDOM.render(
        <NewsApiContext.Provider value={{ listAllNews: () => articles }}>
          <ListAllNews />
        </NewsApiContext.Provider>,
        domElement
      );
    });
    expect(
      Array.from(domElement.querySelectorAll("h3")).map((e) => e.innerHTML)
    ).toEqual([
      "Article 1",
      "Category: Topic 1",
      "Article 2",
      "Category: Topic 2",
    ]);
    expect(domElement.innerHTML).toMatchSnapshot();
  });
});
