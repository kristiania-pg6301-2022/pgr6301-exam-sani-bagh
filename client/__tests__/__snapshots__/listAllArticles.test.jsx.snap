// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListAllNews component should list all news 1`] = `"<div><div><ul><div class=\\"article\\"><h3>Article 1</h3><h3>Category: Topic 1</h3><div></div><p></p></div><div class=\\"article\\"><h3>Article 2</h3><h3>Category: Topic 2</h3><div></div><p></p></div></ul></div></div>"`;

exports[`ListAllNews component should show error message 1`] = `"<div><h1>Error</h1><div id=\\"error-text\\">Error: Failed to load</div></div>"`;

exports[`ListAllNews component should show loading screen 1`] = `"<div>Loading...</div>"`;
